File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/3082/BusinessApplication/Helpers/DataFieldExtensions.vb
Imports System
Imports System.Runtime.CompilerServices
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data

Public Module DataFieldExtensions
    ''' <summary>
    ''' Sustituye un control <see cref="TextBox" /> de <see cref="DataField" /> por otro control y actualiza los enlaces.
    ''' </summary>
    ''' <param name="field"><see cref="DataField"/> cuyo <see cref="TextBox"/> se va a sustituir.</param>
    ''' <param name="newControl">Nuevo control que va a establecer como <see cref="DataField.Content" />.</param>
    ''' <param name="dataBindingProperty">Propiedad del control que se utilizará para el enlace de datos.</param>        
    ''' <param name="bindingSetupFunction">
    ''' <c>Action</c> opcional que se puede utilizar para cambiar parámetros en el enlace recién generado antes de aplicarlo a <paramref name="newControl"/>.
    ''' </param>
    <Extension()> _
    Public Sub ReplaceTextBox(ByVal field As DataField, ByVal newControl As FrameworkElement, ByVal dataBindingProperty As DependencyProperty, _
                              Optional ByVal bindingSetupFunction As Action(Of Binding) = Nothing)
        If field Is Nothing Then
            Throw New ArgumentNullException("field")
        End If

        If newControl Is Nothing Then
            Throw New ArgumentNullException("newControl")
        End If

        ' Construir nuevo enlace al copiar el existente y pasarlo a una función proporcionada por el usuario (si la hubiera)
        Dim newBinding As Binding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy()

        If bindingSetupFunction IsNot Nothing Then
            bindingSetupFunction(newBinding)
        End If

        ' Sustituir campo
        newControl.SetBinding(dataBindingProperty, newBinding)
        field.Content = newControl
    End Sub

    ''' <summary>
    ''' Crea un nuevo objeto <see cref="Binding"/> al copiar todas las propiedades de otro objeto <see cref="Binding"/>.
    ''' </summary>
    ''' <param name="binding"><see cref="Binding"/> del que se copiarán los valores de las propiedades</param>
    ''' <returns>Nuevo objeto <see cref="Binding"/>.</returns>
    <Extension()> _
    Public Function CreateCopy(ByVal binding As Binding) As binding
        If binding Is Nothing Then
            Throw New ArgumentNullException("binding")
        End If

        Dim newBinding As New Binding()

        newBinding.BindsDirectlyToSource = binding.BindsDirectlyToSource
        newBinding.Converter = binding.Converter
        newBinding.ConverterParameter = binding.ConverterParameter
        newBinding.ConverterCulture = binding.ConverterCulture
        newBinding.Mode = binding.Mode
        newBinding.NotifyOnValidationError = binding.NotifyOnValidationError
        newBinding.Path = binding.Path
        newBinding.UpdateSourceTrigger = binding.UpdateSourceTrigger
        newBinding.ValidatesOnExceptions = binding.ValidatesOnExceptions

        If binding.ElementName IsNot Nothing Then
            newBinding.ElementName = binding.ElementName
        ElseIf binding.RelativeSource IsNot Nothing Then
            newBinding.RelativeSource = binding.RelativeSource
        Else
            newBinding.Source = binding.Source
        End If

        Return newBinding
    End Function
End Module

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/2052/BusinessApplication/Helpers/DataFieldExtensions.cs
namespace $safeprojectname$
{
    using System;
    using System.Windows;
    using System.Windows.Controls;
    using System.Windows.Data;

    /// <summary>
    ///     提供扩展方法以对 <see cref="DataField"/> 执行操作。
    /// </summary>
    public static class DataFieldExtensions
    {
        /// <summary>
        /// 将 <see cref="DataField" /> 的 <see cref="TextBox" /> 控件替换为另一个控件并更新绑定。
        /// </summary>
        /// <param name="field">将替换其 <see cref="TextBox"/> 的 <see cref="DataField"/>。</param>
        /// <param name="newControl">将设置为 <see cref="DataField.Content" /> 的新控件。</param>
        /// <param name="dataBindingProperty">将用于数据绑定的控件属性。</param>        
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty)
        {
            field.ReplaceTextBox(newControl, dataBindingProperty, null);
        }

        /// <summary>
        /// 将 <see cref="DataField" /> 的 <see cref="TextBox" /> 控件替换为另一个控件并更新绑定。
        /// </summary>
        /// <param name="field">将替换其 <see cref="TextBox"/> 的 <see cref="DataField"/>。</param>
        /// <param name="newControl">将设置为 <see cref="DataField.Content" /> 的新控件。</param>
        /// <param name="dataBindingProperty">将用于数据绑定的控件属性。</param>        
        /// <param name="bindingSetupFunction">
        /// 一个可选 <see cref="Action"/>，可以用于在将新生成的绑定应用于 <paramref name="newControl"/> 之前更改该绑定的参数
        /// </param>
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty, Action<Binding> bindingSetupFunction)
        {
            if (field == null)
            {
                throw new ArgumentNullException("field");
            }

            if (newControl == null)
            {
                throw new ArgumentNullException("newControl");
            }

            // 通过复制现有绑定并将其发送给 bindingSetupFunction，为调用方要执行的任何更改构造新绑定。
            Binding newBinding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy();

            if (bindingSetupFunction != null)
            {
                bindingSetupFunction(newBinding);
            }

            // 替换字段
            newControl.SetBinding(dataBindingProperty, newBinding);
            field.Content = newControl;
        }

        /// <summary>
        /// 通过从另一个 <see cref="Binding"/> 对象复制所有属性来创建新 <see cref="Binding"/> 对象。
        /// </summary>
        /// <param name="binding">将从其复制属性值的 <see cref="Binding"/></param>
        /// <returns>新 <see cref="Binding"/> 对象。</returns>
        private static Binding CreateCopy(this Binding binding)
        {
            if (binding == null)
            {
                throw new ArgumentNullException("binding");
            }

            Binding newBinding = new Binding()
            {
                BindsDirectlyToSource = binding.BindsDirectlyToSource,
                Converter = binding.Converter,
                ConverterParameter = binding.ConverterParameter,
                ConverterCulture = binding.ConverterCulture,
                Mode = binding.Mode,
                NotifyOnValidationError = binding.NotifyOnValidationError,
                Path = binding.Path,
                UpdateSourceTrigger = binding.UpdateSourceTrigger,
                ValidatesOnExceptions = binding.ValidatesOnExceptions
            };

            if (binding.ElementName != null)
            {
                newBinding.ElementName = binding.ElementName;
            }
            else if (binding.RelativeSource != null)
            {
                newBinding.RelativeSource = binding.RelativeSource;
            }
            else
            {
                newBinding.Source = binding.Source;
            }

            return newBinding;
        }
    }
}
